src/quickgui/plugin/qgsquickplugin.cpp

```diff
@@ -0,0 +1,87 @@
+ ***************************************************************************/
+
+#include <qqml.h>
+
+#include "qgsfeature.h"
+#include "qgsmaplayer.h"
+#include "qgsmessagelog.h"
+#include "qgspointxy.h"
+#include "qgsproject.h"
+#include "qgsrelationmanager.h"
+#include "qgscoordinatetransformcontext.h"
+#include "qgsvectorlayer.h"
+
+#include "qgsquickattributeformmodel.h"
+#include "qgsquickattributeformmodelbase.h"
+#include "qgsquickcoordinatetransformer.h"
+#include "qgsquickfeaturemodel.h"
+#include "qgsquickfeaturehighlight.h"
+#include "qgsquickidentifykit.h"
+#include "qgsquickfeature.h"
+#include "qgsquickmapcanvasmap.h"
+#include "qgsquickmapsettings.h"
+#include "qgsquickmaptransform.h"
+#include "qgsquickmessagelogmodel.h"
+#include "qgsquickplugin.h"
+#include "qgsquickpositionkit.h"
+#include "qgsquickscalebarkit.h"
+#include "qgsquicksubmodel.h"
+#include "qgsquickutils.h"
+
+static QObject *_utilsProvider( QQmlEngine *engine, QJSEngine *scriptEngine )
+{
+  Q_UNUSED( engine )
+  Q_UNUSED( scriptEngine )
+  return QgsQuickUtils::instance();  // the object will be owned by QML engine and destroyed by the engine on exit
+}
+
+void QgisQuickPlugin::registerTypes( const char *uri )
+{
+  qDebug( "REGISTERING QQmlExtensionInterface: QgisQuick" );
+
+  qRegisterMetaType< QList<QgsMapLayer *> >( "QList<QgsMapLayer*>" );
+  qRegisterMetaType< QgsAttributes > ( "QgsAttributes" );
+  qRegisterMetaType< QgsCoordinateReferenceSystem >( "QgsCoordinateReferenceSystem" );
+  qRegisterMetaType< QgsCoordinateTransformContext >( "QgsCoordinateTransformContext" );
+  qRegisterMetaType< QgsFeature > ( "QgsFeature " );
+  qRegisterMetaType< QgsFeatureId > ( "QgsFeatureId" );
+  qRegisterMetaType< QgsPoint >( "QgsPoint" );
+  qRegisterMetaType< QgsPointXY >( "QgsPointXY" );
+  qRegisterMetaType< QgsQuickFeature >( "QgsQuickFeature" );
+
+  qmlRegisterType< QgsProject >( uri, 0, 1, "Project" );
+  qmlRegisterType< QgsQuickAttributeFormModel >( uri, 0, 1, "AttributeFormModel" );
+  qmlRegisterType< QgsQuickCoordinateTransformer >( uri, 0, 1, "CoordinateTransformer" );
+  qmlRegisterType< QgsQuickFeatureModel >( uri, 0, 1, "FeatureModel" );
+  qmlRegisterType< QgsQuickFeatureHighlight >( uri, 0, 1, "FeatureHighlight" );
+  qmlRegisterType< QgsQuickIdentifyKit >( uri, 0, 1, "IdentifyKit" );
+  qmlRegisterType< QgsQuickMapCanvasMap >( uri, 0, 1, "MapCanvasMap" );
+  qmlRegisterType< QgsQuickMapSettings >( uri, 0, 1, "MapSettings" );
+  qmlRegisterType< QgsQuickMapTransform >( uri, 0, 1, "MapTransform" );
+  qmlRegisterType< QgsQuickMessageLogModel >( uri, 0, 1, "MessageLogModel" );
+  qmlRegisterType< QgsQuickPositionKit >( uri, 0, 1, "PositionKit" );
+  qmlRegisterType< QgsQuickScaleBarKit >( uri, 0, 1, "ScaleBarKit" );
+  qmlRegisterType< QgsQuickSubModel >( uri, 0, 1, "SubModel" );
+  qmlRegisterType< QgsRelationManager >( uri, 0, 1, "RelationManager" );
+  qmlRegisterType< QgsVectorLayer >( uri, 0, 1, "VectorLayer" );
```

Comment:
Should "native" types like QgsVectorLayer etc. be exported from the same module as the QgisQuick types?
Should we use them unprefixed (VectorLayer rather than QgsVectorLayer)? Is it obvious for a developer that the API documentation of QgsVectorLayer applies in this case?

Reply:
I do not have a strong opinion on any of these. But

> Should "native" types like QgsVectorLayer etc. be exported from the same module as the QgisQuick types?

It is better if you want to overwrite QGIS Core class, e.g. see QgsQuickMapSettings: public QgsMapSettings. So you do not need to change the application code. 

> Should we use them unprefixed (VectorLayer rather than QgsVectorLayer)?

We decided based on QT QML class names, e.g. in QML you use `Item`, but C++ class is `QQuickItem`

> Is it obvious for a developer that the API documentation of QgsVectorLayer applies in this case?

Not sure, probably using the full name would help google it.